### PR TITLE
run notebooks after tests as part of CI

### PR DESCRIPTION
**Pull request by orm011**
Adds a check that the notebooks don't fail to run against the latest changes.
Let's wait and see if CI yaml still works ok.

There is separate work we need to do to keep the HTML in the docs up to date to the latest code. 
This needs to happen before the doc generating step, on that pipeline.

- - -
This change is [Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/109)

### CODE REVIEW

File: .github/workflows/pytest.yml [hunk 1]
```diff
@@ -45,3 +45,12 @@ jobs:
           FIREWORKS_API_KEY: ${{ secrets.FIREWORKS_API_KEY }}
           OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
           TOGETHER_API_KEY: ${{ secrets.TOGETHER_API_KEY }}
+      - name: Run the documentation notebooks
+        run: poetry run pytest -v --nbmake ./docs/**/*ipynb
```
Comment (aaron-siegel) [inline .github/workflows/pytest.yml]: Are you sure this is the right command? When I run it locally all the notebooks fail with errors like

`AttributeError: module 'pixeltable' has no attribute 'functions'`

I know that rag-demo, at least, should be working, and if not I wouldn't expect this to be the error?

**Comment by aaron-siegel** [comment]
Thought about this a bit. Before it's ready for production, I'd suggest it needs:
- To be working on at least one notebook (we don't have to fix all the notebooks as part of this PR, but at least one of them should be working so we know the environment is set up properly)
- Instructions for how to run it locally successfully (so we can preview test results before pushing to CI)
- A story around how we isolate the environment for running notebook tests locally. Right now (if I'm not mistaken) the notebook tests will run against the user's local "production" DB - unlike other unit tests, which run against a separate test DB instance. We should figure out how to configure the notebook tests to run against the test instance.

**Comment by orm011** [comment]
*[`.github/workflows/pytest.yml` line 49 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/109#gh-1544972975:-NuPYGky5k5o6dGEetvu:b-n7p0mk) ([raw file](https://github.com/pixeltable/pixeltable/blob/82acd0c7fdbb7972d580d29ed564c2dd90dc7616/.github/workflows/pytest.yml#L49)):*
<details><summary><i>Previously, orm011 (Oscar Moll) wrote…</i></summary><blockquote>

There's some larger environment issue orthogonal to  this PR. 
I built a wheel for the project using poetry build, and then installed that wheel on a clean conda with only python 3.9 .

The same error occurs (in addition to a few missing dependencies that I had to adjust prior to running into it, but also outside of this PR)
</blockquote></details>

I think the odd one out is poetry run pytest, which does work on the dev env.. But I can reproduce this error even in dev env by doing cl = pxt.Client() in Ipython

**Review by orm011 — COMMENTED** [review]
I've added changes so  CIs runs files without getpass on PR, and  CI runs all tests on push. Nothing will break CI for now, because we need to see what/how things fail in that github runner env.  

> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/109#-:-NygPpGb7fl9CWw_hWBS:b58nbhs)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @aaron-siegel)

**Review by aaron-siegel — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/109)* status: 0 of 8 files reviewed, 5 unresolved discussions (waiting on @orm011)

___
*[`Makefile` line 78 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/109#-NygRw6YBaFurIIbqZ1F:-NygRw6YBaFurIIbqZ1G:b-3b30an) ([raw file](https://github.com/pixeltable/pixeltable/blob/9682a53a4f471e867a141aaf4cc67f35dfb85fb4/Makefile#L78)):*
> ```Makefile
> AUTH_FREE_NBS := $(shell grep -L "getpass"  docs/release/**/*ipynb)
> .PHONY: auth-free-notebooks
> auth-free-notebooks: $(AUTH_FREE_NBS)
> ```

I'd prefer this be controlled by an environment variable rather than a separate maketarget.

I'm envisioning an environment variable $(NB_MAKE_TARGETS) or some such, that defaults to `docs/release/**/*.ipynb` if not set.

Actually, thinking about this some more, see the below comment about running `make` in CI. If we follow that course, it kind of obviates this comment.
___
*[`.github/workflows/pytest.yml` line 49 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/109#gh-1544972975:-NygSEmRArQx55RkBi5_:b-qp2xic) ([raw file](https://github.com/pixeltable/pixeltable/blob/82acd0c7fdbb7972d580d29ed564c2dd90dc7616/.github/workflows/pytest.yml#L49)):*
<details><summary><i>Previously, orm011 (Oscar Moll) wrote…</i></summary><blockquote>

I think the odd one out is poetry run pytest, which does work on the dev env.. But I can reproduce this error even in dev env by doing cl = pxt.Client() in Ipython
</blockquote></details>

`poetry run` should not be necessary anymore.
___
*[`.github/workflows/pytest.yml` line 88 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/109#-NygTD1G8lbDLb6eDQlS:-NygTD1G8lbDLb6eDQlT:b4be413) ([raw file](https://github.com/pixeltable/pixeltable/blob/9682a53a4f471e867a141aaf4cc67f35dfb85fb4/.github/workflows/pytest.yml#L88)):*
> ```YAML
>         if: ${{ !startsWith(github.ref, 'refs/pull/') }}
>         run: make notebooks
>         continue-on-error: true # For now, continue even if the notebooks fail
> ```

Two spaces before a comment, please (following PEP8 style: https://peps.python.org/pep-0008/#inline-comments)
___
*[`.github/workflows/pytest.yml` line 90 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/109#-NygT8xt51RUk0QfxPLv:-NygT8xt51RUk0QfxPLw:beabs5q) ([raw file](https://github.com/pixeltable/pixeltable/blob/9682a53a4f471e867a141aaf4cc67f35dfb85fb4/.github/workflows/pytest.yml#L90)):*
> ```YAML
>         continue-on-error: true # For now, continue even if the notebooks fail
>         env:
>           FIREWORKS_API_KEY: ${{ secrets.FIREWORKS_API_KEY }}
> ```

Factor these out into global env so that we're not repeating ourselves in the makefile (this list will surely grow)
___
*[`.github/workflows/pytest.yml` line 95 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/109#-NygSPVs7s75OQKB6zaT:-NygSPVs7s75OQKB6zaU:bdmorb1) ([raw file](https://github.com/pixeltable/pixeltable/blob/9682a53a4f471e867a141aaf4cc67f35dfb85fb4/.github/workflows/pytest.yml#L95)):*
> ```YAML
>       - name: Run all notebooks that do not require secret keys
>         if: ${{ startsWith(github.ref, 'refs/pull/') }}
>         run: make auth-free-notebooks
> ```

Now that I think about it, we should probably avoid running `make` at all in CI. It's really intended for a dev environment. Why don't we just run the pytest command directly in CI? Then we can easily control the selection of notebooks, within CI.

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/109)* status: 0 of 8 files reviewed, 5 unresolved discussions (waiting on @aaron-siegel)

___
*[`Makefile` line 78 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/109#-NygRw6YBaFurIIbqZ1F:-NygdKKCBkLjE2cie4rq:bylj1r4) ([raw file](https://github.com/pixeltable/pixeltable/blob/9682a53a4f471e867a141aaf4cc67f35dfb85fb4/Makefile#L78)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

I'd prefer this be controlled by an environment variable rather than a separate maketarget.

I'm envisioning an environment variable $(NB_MAKE_TARGETS) or some such, that defaults to `docs/release/**/*.ipynb` if not set.

Actually, thinking about this some more, see the below comment about running `make` in CI. If we follow that course, it kind of obviates this comment.
</blockquote></details>

Done. followed advice below
___
*[`.github/workflows/pytest.yml` line 49 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/109#gh-1544972975:-NygdU3T9bDDAoUJhGm-:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/82acd0c7fdbb7972d580d29ed564c2dd90dc7616/.github/workflows/pytest.yml#L49)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

`poetry run` should not be necessary anymore.
</blockquote></details>

Done.
___
*[`.github/workflows/pytest.yml` line 88 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/109#-NygTD1G8lbDLb6eDQlS:-NygdjHp0VJvayQtgY6M:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/9682a53a4f471e867a141aaf4cc67f35dfb85fb4/.github/workflows/pytest.yml#L88)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

Two spaces before a comment, please (following PEP8 style: https://peps.python.org/pep-0008/#inline-comments)
</blockquote></details>

Done.
___
*[`.github/workflows/pytest.yml` line 90 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/109#-NygT8xt51RUk0QfxPLv:-NygdjvzEf6qqcEgK-VC:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/9682a53a4f471e867a141aaf4cc67f35dfb85fb4/.github/workflows/pytest.yml#L90)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

Factor these out into global env so that we're not repeating ourselves in the makefile (this list will surely grow)
</blockquote></details>

Done.
___
*[`.github/workflows/pytest.yml` line 95 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/109#-NygSPVs7s75OQKB6zaT:-Nygdkuu8jG5A9Qs6nNB:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/9682a53a4f471e867a141aaf4cc67f35dfb85fb4/.github/workflows/pytest.yml#L95)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

Now that I think about it, we should probably avoid running `make` at all in CI. It's really intended for a dev environment. Why don't we just run the pytest command directly in CI? Then we can easily control the selection of notebooks, within CI.
</blockquote></details>

Done.

**Review by aaron-siegel — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/109)* status: 0 of 8 files reviewed, 3 unresolved discussions (waiting on @orm011)

___
*[`.github/workflows/pytest.yml` line 92 at r5](https://reviewable.io/reviews/pixeltable/pixeltable/109#-Nygm2BZEMpsho4DEirN:-Nygm2BZEMpsho4DEirO:b-myx98) ([raw file](https://github.com/pixeltable/pixeltable/blob/a31725e729b6c29af2d90c9554cd39a67bcde371/.github/workflows/pytest.yml#L92)):*
> ```YAML
>           GITHUB_REF: ${{ github.ref }}
>         run: |
>           if [[ $GITHUB_REF == refs/pull/* ]]; then
> ```

Best practice in bash scripts is to quote the variable here:

```
if [[ "$GITHUB_REF" == refs/pull/* ]]; then
```
It avoids strange errors when the variable contains spaces (which we don't expect it to, but... quoting the variable has just become automatic for me)
___
*[`docs/release/tutorials/rag-operations.ipynb` line 1966 at r5](https://reviewable.io/reviews/pixeltable/pixeltable/109#-NygmiJA5cNBfY3dsOSX:-NygmiJA5cNBfY3dsOSY:b-n9bxff) ([raw file](https://github.com/pixeltable/pixeltable/blob/a31725e729b6c29af2d90c9554cd39a67bcde371/docs/release/tutorials/rag-operations.ipynb#L1966)):*
> ```
>      "text": [
>       "\r",
>       "Computing cells:  23%|█████████                              | 112/480 [00:00<00:02, 164.02 cells/s]"
> ```

Any idea what's happening here? It's going to look really weird. With the manually executed notebooks there was just a single line of output here.

**Review by aaron-siegel — APPROVED** [review]

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/109)* status: 0 of 9 files reviewed, 3 unresolved discussions (waiting on @aaron-siegel)

___
*[`.github/workflows/pytest.yml` line 92 at r5](https://reviewable.io/reviews/pixeltable/pixeltable/109#-Nygm2BZEMpsho4DEirN:-NyldOmU21zQU72s-WwP:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/a31725e729b6c29af2d90c9554cd39a67bcde371/.github/workflows/pytest.yml#L92)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

Best practice in bash scripts is to quote the variable here:

```
if [[ "$GITHUB_REF" == refs/pull/* ]]; then
```
It avoids strange errors when the variable contains spaces (which we don't expect it to, but... quoting the variable has just become automatic for me)
</blockquote></details>

Done.
___
*[`docs/release/tutorials/rag-operations.ipynb` line 1966 at r5](https://reviewable.io/reviews/pixeltable/pixeltable/109#-NygmiJA5cNBfY3dsOSX:-NyldRRW3jHfffP88Ia4:b-72k4xn) ([raw file](https://github.com/pixeltable/pixeltable/blob/a31725e729b6c29af2d90c9554cd39a67bcde371/docs/release/tutorials/rag-operations.ipynb#L1966)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

Any idea what's happening here? It's going to look really weird. With the manually executed notebooks there was just a single line of output here.
</blockquote></details>

will investigate... for now just pushing notebooks with clean stuff. jupyter-nbconvert, runs notebooks for you and saves the new outputs has the same issue.

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/109)* status: 0 of 10 files reviewed, 3 unresolved discussions (waiting on @aaron-siegel)

___
*[`docs/release/tutorials/rag-operations.ipynb` line 1966 at r5](https://reviewable.io/reviews/pixeltable/pixeltable/109#-NygmiJA5cNBfY3dsOSX:-NylsOOBD9jv197j22oK:b-1bj03u) ([raw file](https://github.com/pixeltable/pixeltable/blob/a31725e729b6c29af2d90c9554cd39a67bcde371/docs/release/tutorials/rag-operations.ipynb#L1966)):*
<details><summary><i>Previously, orm011 (Oscar Moll) wrote…</i></summary><blockquote>

will investigate... for now just pushing notebooks with clean stuff. jupyter-nbconvert, runs notebooks for you and saves the new outputs has the same issue.
</blockquote></details>

fixed via env var in makefile

**Pull request merged**